src/evm_circuit/arithmetic/add.rs

```diff
@@ -0,0 +1,353 @@
+                region.assign_advice(
+                    || format!("carry {}", idx),
+                    self.word_config.bytes[idx],
+                    offset,
+                    || {
+                        carry
+                            .map(|carry| F::from_u64(carry as u64))
+                            .ok_or(Error::SynthesisError)
+                    },
+                )
+            })
+            .collect();
+        let offset = offset + 1;
+
+        let c = self.word_config.assign_word(region, offset, c)?;
+        let offset = offset + 1;
+
+        region.assign_fixed(|| "q_group_4", self.q_group_4, offset, || Ok(F::one()))?;
+
+        Ok((a, b, c))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::gadget::evm_word;
+    use halo2::{
+        circuit::{Layouter, SimpleFloorPlanner},
+        dev::MockProver,
+    };
+    use std::{convert::TryInto, marker::PhantomData};
+
+    // Note: Using the MockProver here causes an out-of-memory error.
+    use halo2::plonk::*;
+    use pasta_curves::{arithmetic::FieldExt, pallas};
+
+    #[test]
+    fn test_add() {
+        #[derive(Default)]
+        struct AddCircuit<F: FieldExt> {
+            a: [Option<u8>; 32],
+            b: [Option<u8>; 32],
+            c: [Option<u8>; 32],
+            carry: [Option<bool>; 32],
+            _marker: PhantomData<F>,
+        }
+
+        impl<F: FieldExt> Circuit<F> for AddCircuit<F> {
+            type Config = Config<F>;
+            type FloorPlanner = SimpleFloorPlanner;
+
+            fn without_witnesses(&self) -> Self {
+                Self::default()
+            }
+
+            fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+                let q_group_4 = meta.fixed_column();
+                let op = meta.advice_column();
+                let global_counter = meta.advice_column();
+                let stack_pointer = meta.advice_column();
+                let bytes = (0..32).map(|_| meta.advice_column()).collect::<Vec<_>>();
+
+                Config::configure(
+                    evm_word::r(),
+                    meta,
+                    q_group_4,
+                    op,
+                    global_counter,
+                    stack_pointer,
+                    bytes.try_into().unwrap(),
+                )
+            }
+
+            fn synthesize(
+                &self,
+                config: Self::Config,
+                mut layouter: impl Layouter<F>,
+            ) -> Result<(), Error> {
+                config.word_config.load(&mut layouter)?;
+
+                layouter.assign_region(
+                    || "a + b = c",
+                    |mut region| config.assign(&mut region, 0, self.a, self.b, self.c, self.carry),
+                )?;
+
+                Ok(())
+            }
+        }
+
+        let a = pallas::Base::rand().to_bytes();
+        let b = pallas::Base::rand().to_bytes();
+        let mut c: Vec<u8> = Vec::with_capacity(32);
```

Comment:
unnecessary annotation

```suggestion
        let mut c = Vec::with_capacity(32);
```